File: apps/ios/0dteTrader/Core/Networking/APIError.swift
import Foundation

enum APIError: Error, Equatable {
    /// Server responded with `{ "error": { "code", "message" } }`.
    case server(code: String, message: String, status: Int)
    /// Non-2xx response without a decodable error envelope.
    case httpStatus(Int)
    /// Transport-level failure (no connectivity, timeout, ...).
    case network(underlying: String)
    /// Response body could not be decoded into the expected type.
    case decoding
    /// The request could not even be constructed.
    case invalidRequest
    /// Access token expired and the refresh token was rejected; the session is dead.
    case unauthorized

    var userMessage: String {
        switch self {
        case let .server(_, message, _):
            return message
        case let .httpStatus(status):
            return "Request failed (HTTP \(status))."
        case let .network(underlying):
            return "Network error: \(underlying)"
        case .decoding:
            return "Unexpected response from server."
        case .invalidRequest:
            return "Invalid request."
        case .unauthorized:
            return "Session expired. Please log in again."
        }
    }
}

File: apps/desktop/src/core/api/ApiError.ts
import type { ApiError as ApiErrorEnvelope } from '@0dtetrader/shared-types';

/** Typed API failure, mirroring APIError.swift including the user messages. */
export type ApiFailure =
  | { kind: 'server'; code: string; message: string; status: number }
  | { kind: 'httpStatus'; status: number }
  | { kind: 'network'; underlying: string }
  | { kind: 'decoding' }
  | { kind: 'invalidRequest' }
  | { kind: 'unauthorized' };

export class ApiError extends Error {
  readonly failure: ApiFailure;

  constructor(failure: ApiFailure) {
    super(userMessage(failure));
    this.name = 'ApiError';
    this.failure = failure;
  }

  get userMessage(): string {
    return userMessage(this.failure);
  }

  static isUnauthorized(error: unknown): boolean {
    return error instanceof ApiError && error.failure.kind === 'unauthorized';
  }
}

function userMessage(failure: ApiFailure): string {
  switch (failure.kind) {
    case 'server':
      return failure.message;
    case 'httpStatus':
      return `Request failed (HTTP ${failure.status}).`;
    case 'network':
      return `Network error: ${failure.underlying}`;
    case 'decoding':
      return 'Unexpected response from server.';
    case 'invalidRequest':
      return 'Invalid request.';
    case 'unauthorized':
      return 'Session expired. Please log in again.';
  }
}

/** Error message shown to the user for any thrown value. */
export function errorMessage(error: unknown): string {
  if (error instanceof ApiError) return error.userMessage;
  if (error instanceof Error) return error.message;
  return String(error);
}

export function parseErrorEnvelope(body: unknown): { code: string; message: string } | null {
  const envelope = body as Partial<ApiErrorEnvelope> | null;
  if (
    envelope &&
    typeof envelope === 'object' &&
    envelope.error &&
    typeof envelope.error.code === 'string' &&
    typeof envelope.error.message === 'string'
  ) {
    return envelope.error;
  }
  return null;
}
